### ISSUE TITLE
Enhancement: [switch-exhaustiveness-check] Turn on requireDefaultForNonUnion by default

### ISSUE THREAD

**Issue by ST-DDT** [typescript-eslint/typescript-eslint#7966]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/switch-exhaustiveness-check/

### Description

This issue is there to discuss whether the `requireDefaultForNonUnion` **option** should be enabled by default.
This issue is **not** about whether the `switch-exhaustiveness-check` should be part of any preset or the option to be removed entirely.

### Fail

```typescript
declare const value: number;
switch (value) {
  case 0:
    return 0;
  case 1:
    return 1;
}
```

### Pass

```typescript
declare const value: number;
switch (value) {
  case 0:
    return 0;
  case 1:
    return 1;
  default:
    return -1;
}
```

### Additional Info

Based on https://github.com/typescript-eslint/typescript-eslint/pull/7880#issuecomment-1819547893

- https://github.com/typescript-eslint/typescript-eslint/pull/7880#issuecomment-1819547893

Partial previous discussion: https://github.com/typescript-eslint/typescript-eslint/issues/2959

- https://github.com/typescript-eslint/typescript-eslint/issues/2959

**Comment by ST-DDT** [comment]
IMO you always want to cover all cases in your switch, either by exhausting all potential cases, or by having a default statement.
If you don't think, that exhausting all potential cases is the right way to go, then why would you enable the rule in the first place?

If a type doesn't have type information or overly broad type information (such as number) and don't have a default case, then you are either implicitly casting the type to more specific type
or implicitly hide the "else do nothing" branch and thus lead future readers of the code astray, that only these cases exist (especially if all of your other switches are exhaustive).

I did a non-exhaustive-check on this repository for `switch` statements and they either had an implicit default (by having a return like statement directly after the switch) or an explicit default.

[curious]: Can somebody please point me to a real piece of code/repository, that has a non-exhaustive-switch on a non-union-type, but has otherwise most of their union-type switches exhaustive?
(I know plenty repositories, that do or do not care about exhaustiveness, but I don't know any that decide that on the type of the value)

**Comment by ArnaudBarre** [comment]
Thanks for creating the issue to discuss this. My personal opinion is that the option should be removed. I'm pretty sure that if the rule was created first with this check nobody will have ask to create this intermediate check.

**Comment by Josh-Cena** [comment]
We cannot cause extra errors in minor versions, unless the error is highly expectable. People may not expect non-union types to be checked at all. Of course, unless other team members disagree with me, but I think this is sufficiently debatable that it can only be enabled by default (or held up for removal) in the next major version.

**Comment by ArnaudBarre** [comment]
Yeah I know that this is the current consensus that updates should not help you catch new bugs unless you read carefully the changelog to enable new flags, I'm just giving my opinion by phrasing it this way.

**Comment by JoshuaKGoldberg** [comment]
Yup happy to talk about this as a breaking change!

My main concern here is that doing so is very strict for a case that doesn't _directly_ catch bugs. It just often _indirectly_ catches them (which is why folks are in favor of it existing!). We (linters) have learned the hard way that if you add a behavior that "just" restricts some code pattern without provably showing the bug it fixed, that'll annoy folks. You need to really be able to show that it's >99% of the time _definitely_ useful _immediately_. Again, not saying this issue's proposal is wrong, just that we've learned it's very hard to put changes like this in front of people.

Trying this out in the `typescript-eslint` repo by enabling `requireDefaultForNonUnion` in our root-level ESLint config, we do see a few cases where the option doesn't make sense. Looking at the first two:

* A `switch` over a `string`, where there are a couple of `case`s for known special casing characters: https://github.com/typescript-eslint/typescript-eslint/blob/09c04fb0d5a5dee0dc266abb2e21aa9a1c489712/packages/ast-spec/tests/util/serializers/string.ts#L21
* A `switch` over a child node in AST analysis, where we want to run some logic only for a couple kinds of child nodes: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/scope-manager/src/referencer/Referencer.ts

In both those cases the code is working fine as-is. There's nothing _directly_ to be gained from adding a `default` case. One could argue that always having a `default` in a non-exhaustive `switch` is helpful stylistically... but that's a stylistic opinion that not everyone would have. I would go for the opposite: less code.

Marking as `evaluating community engagement` to see if anybody has strong counterpoints here. But unless there's a big swell of user interest, I don't see the arguments in favor of the `requireDefaultForNonUnion ` option _existing_ being powerful enough to justify it being on by default.

> I did a non-exhaustive-check on this repository for switch statements and they either had an implicit default (by having a return like statement directly after the switch) or an explicit default.

Aside: I _love_ it when people look at the data for suggestions. This is very helpful to know 😄 thanks!

**Comment by ST-DDT** [comment]
I see why you consider this stylistic. IMO both examples are bad switches because they only operate on a very tiny subset of the available data and might take just as much characters as written as an if statement (and then they wouldn't need a comment regarding their codeflow). But I see why you don't want to add extra characters to any stylistic choice.

> Aside: I _love_ it when people look at the data for suggestions. This is very helpful to know 😄 thanks!

I hope this gets easier/less time consuming/less patchy with AI in the future.

**Comment by ArnaudBarre** [comment]
Yeah ok I see and i agree that when you write switch in place of `else if` it doesn't really lead to bugs to omit the default. So I understand the choice to not enable it by default.

**Comment by JoshuaKGoldberg** [comment]
It's been ~4 months since last comment and there haven't been any new motivators for this. Closing it out as having not had enough community feedback to suggest we should take action.

If anybody sees this and thinks they have a good reason not yet mentioned, please do file a new issue! We'd be happy to take a look.

Thanks for the discussion all!
